src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+          case (#ok) lastChunkTimestamp := Time.now();
+        };
+        queue.pruneTo(headId + elements.size());
+        restoreHistoryIfNeeded();
+        #ok(elements.size());
+      } catch (err : Error) {
+        concurrentChunksCounter -= 1;
+        lowestError := #Val(switch (lowestError) { case (#Inf) headId; case (#Val val) Nat.min(val, headId) });
+        restoreHistoryIfNeeded();
+        #err(#SendChunkError(Error.message(err)));
+      };
+    };
+
+    private func restoreHistoryIfNeeded() {
+      switch (lowestError) {
+        case (#Inf) {};
+        case (#Val val) {
+          if (val < queue.rewindIndex()) { Debug.trap("cannot happen") };
+          if (val == queue.rewindIndex()) {
+            queue.rewind();
+            lowestError := #Inf;
+          };
+        };
+      };
+    };
+
+    class StreamIter<T>(queue : QueueBuffer.QueueBuffer<T>, weightFunc : (item : T) -> Nat) {
+      var remainingWeight = 0;
+      public func reset() : () {
+        remainingWeight := weightLimit;
+      };
+      public func next() : ?T {
+        if (remainingWeight == 0) { return null };
+        switch (queue.peek()) {
+          case (?(_, el)) {
+            let weight = weightFunc(el);
+            if (remainingWeight < weight) {
+              return null;
+            } else {
+              ignore queue.pop();
+              remainingWeight -= weight;
+              return ?el;
+            };
+          };
+          case _ {} // queue was empty: stop iteration
+        };
+        return null;
+      };
+    };
+    let streamIter = StreamIter<T>(queue, weightFunc);
+  };
+
+  public type StreamSource = { #canister : Principal; #internal };
+
+  public type StreamInfo<T> = {
+    source : StreamSource;
+    var nextItemId : Nat;
+    var receiver : ?StreamReceiver<T>;
+  };
+
+  type StableStreamInfo = {
+    source : StreamSource;
+    nextItemId : Nat;
+    active : Bool;
+  };
+
+  public type ManagerStableData = (Vec.Vector<StableStreamInfo>, Vec.Vector<(Principal, ?Nat)>);
+  public func defaultManagerStableData() : ManagerStableData = (Vec.new(), Vec.new());
+
+  func requireOk<T>(res : R.Result<T, Any>) : T {
+    switch (res) {
+      case (#ok ok) ok;
+      case (_) Debug.trap("Required result is #err");
+    };
+  };
+
+  public class StreamsManager<T>(
+    initialSourceCanisters : [Principal],
+    itemCallback : (streamId : Nat, sourceCanisterIndex : ?Nat, item : T, index : Nat) -> Any,
+  ) {
+
+    // info about each issued stream id is preserved here forever. Index is a stream ID
+    let streams_ : Vec.Vector<StreamInfo<T>> = Vec.new();
+    // a list of registered source canister principals
+    let sourceCanisters_ : Vec.Vector<Principal> = Vec.fromArray(initialSourceCanisters);
+    // a list of source canister current stream id-s, can be mapped to 'sourceCanisters' by indices
+    let sourceCanisterStreamIds_ : Vec.Vector<?Nat> = Vec.init<?Nat>(initialSourceCanisters.size(), null);
```

Comment:
It seems like we are using these two together as a map source principal -> oldStreamId. (going indirectly over "source canister index") But we are doing a linear search anyway, so we are not worried about efficiency (which is ok because the number of sources is small).

But why not use an AssocList then with K = source Principal and V = existing (old) stream id?

Reply:
Then we get deduplication for free (see comment further below).

Reply:
Done